pandas/core/groupby/generic.py

```diff
@@ -396,11 +396,13 @@ def _wrap_applied_output(
             res_ser.name = self.obj.name
             return res_ser
         elif isinstance(values[0], (Series, DataFrame)):
-            return self._concat_objects(
+            result = self._concat_objects(
```

Comment:
I think there are a few lines at the end of _concat_objects in groupby.py that can be removed since you're now setting the name here; can you check?

Reply:
Right now, removing these lines breaks at least one test, the reason being `DataFrameGroupBy.apply` can return a Series without going through a `SeriesGroupBy`. E.g. in
```
df = DataFrame([[1, 2], [3, 4]])
df.groupby(0, as_index=False)[1].apply(lambda x: x)
```
the result loses its name.

It's not clear to me how easy it will be to take care of these kind of cases.

If keeping the lines, we can possibly slightly simplify them, for instance going from 
```
        name = self.obj.name if self.obj.ndim == 1 else self._selection
        if isinstance(result, Series) and name is not None:

            result.name = name
```
to
```
        if isinstance(result, Series) and self._selection is not None:
            result.name = self._selection
```
but i'm not completely sure this doesn't/won't expose to some edge case.

Reply:
Thanks for checking - I forgot that by using `__getitem__` you can have a DataFrameGroupby with self.obj.ndim == 1. Makes sense to leave as-is for now.